File: web/src/server/store.ts
import { readFile, writeFile, mkdir } from "node:fs/promises";
import { dirname, join } from "node:path";
import type { ContribRecord, ContribStore } from "./types";

// The store is a single JSON file under web/data/ (git-ignored). One file is
// enough at this scale (tens–hundreds of memories) per the spec's "SQLite or a
// JSON file". Pure ops below are unit-tested; load/save are the fs seam.
export const STORE_PATH =
  process.env.MEMORIES_STORE_PATH ?? join(process.cwd(), "data", "memories.json");

export function emptyStore(): ContribStore {
  return { records: [] };
}

export function findById(store: ContribStore, id: string): ContribRecord | undefined {
  return store.records.find((r) => r.id === id);
}

export function addRecord(store: ContribStore, record: ContribRecord): ContribStore {
  return { records: [...store.records, record] };
}

export function updateRecord(
  store: ContribStore,
  id: string,
  patch: Partial<ContribRecord>,
): ContribStore {
  return {
    records: store.records.map((r) => (r.id === id ? { ...r, ...patch, id: r.id } : r)),
  };
}

/** fs seam: read the store, tolerating a missing file (first run → empty). */
export async function loadStore(path: string = STORE_PATH): Promise<ContribStore> {
  try {
    const raw = await readFile(path, "utf8");
    const parsed = JSON.parse(raw) as ContribStore;
    return Array.isArray(parsed.records) ? parsed : emptyStore();
  } catch (err) {
    if ((err as NodeJS.ErrnoException).code === "ENOENT") return emptyStore();
    throw err;
  }
}

/** fs seam: write the store, creating web/data/ if needed. */
export async function saveStore(store: ContribStore, path: string = STORE_PATH): Promise<void> {
  await mkdir(dirname(path), { recursive: true });
  await writeFile(path, JSON.stringify(store, null, 2));
}

File: backend/store.py
"""The contribution store: a single JSON file holding every lifecycle state.

Port of ``web/src/server/store.ts``. Pure ops (empty/find/add/update) are unit-tested;
load/save are the filesystem seam. One file is enough at this scale (tens-hundreds of
memories). Records are plain dicts matching the JSON on disk.
"""
from __future__ import annotations

import json
from pathlib import Path
from typing import Optional

from .config import STORE_PATH


def empty_store() -> dict:
    return {"records": []}


def find_by_id(store: dict, record_id: str) -> Optional[dict]:
    for r in store["records"]:
        if r.get("id") == record_id:
            return r
    return None


def add_record(store: dict, record: dict) -> dict:
    """Append immutably (returns a new store; input untouched)."""
    return {"records": [*store["records"], record]}


def update_record(store: dict, record_id: str, patch: dict) -> dict:
    """Patch a single record by id, immutably; no-op for an unknown id. ``id`` is kept."""
    records = []
    for r in store["records"]:
        if r.get("id") == record_id:
            merged = {**r, **patch, "id": r["id"]}
            records.append(merged)
        else:
            records.append(r)
    return {"records": records}


# --- filesystem seam ------------------------------------------------------------

def load_store(path: Path = STORE_PATH) -> dict:
    """Read the store, tolerating a missing file (first run -> empty)."""
    try:
        raw = Path(path).read_text(encoding="utf8")
    except FileNotFoundError:
        return empty_store()
    parsed = json.loads(raw)
    return parsed if isinstance(parsed.get("records"), list) else empty_store()


def save_store(store: dict, path: Path = STORE_PATH) -> None:
    """Write the store, creating the data directory if needed."""
    path = Path(path)
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(json.dumps(store, indent=2), encoding="utf8")
